### PR TITLE
dbld: Fix missing repositories in Fedora rpm workaround

### PR DESCRIPTION
**Pull request by gaborznagy**
**TLDR:**
Some repositories (updates, updates-modular and fedora-modular) were using the same URL as the "fedora" repository, which caused that some package updates were not visible, leading to package conflicts with the existing package installation.

You can check with `dnf repolist --verbose` to see that current mirror (if no baseurl is used in the .repo file), or query the metalink URL.

Also, we should use primarily the metalink service instead of disabling it and only use a specific mirror as fallback mechanism (instead of by default using a deidcated mirror).
I've checked that if `metalink` is set to a non-existing address (see the test script below), `baseurl` is used as a fallback, and as I've seen errors about metalink address couldn't be resolved, it proves that metalink has a higher priority than baseurl.

**Motivation:**
When reviewing the patches of #3933 I've found it strange that `vim` had to be excluded fro installation due to packaging errors.
I've checked a stock fedora 35 docker image, and there were no problem with it.
Then I've checked the dbld image of Fedora 35 and tried to install vim to see the problem.
Turned out the `vim-minimal` pre-installed package had a newer version, while installing `vim` with an old version caused the conflict.

<details>

```
# dnf install -y vim
Copr repo for syslog-ng-githead owned by czanik                                                                                                                                                                                          36 kB/s |  37 kB     00:01    
Fedora 35 - x86_64                                                                                                                                                                                                                       27 kB/s | 6.1 kB     00:00    
Error: Cannot find a valid baseurl for repo: fedora-cisco-openh264
Fedora Modular 35 - x86_64                                                                                                                                                                                                               28 kB/s | 6.1 kB     00:00    
Fedora 35 - x86_64 - Updates                                                                                                                                                                                                             20 kB/s | 6.1 kB     00:00    
Fedora Modular 35 - x86_64 - Updates                                                                                                                                                                                                     12 kB/s | 6.1 kB     00:00    
Ignoring repositories: fedora-cisco-openh264
Dependencies resolved.
========================================================================================================================================================================================================================================================================
 Package                                                            Architecture                                               Version                                                                 Repository                                                  Size
========================================================================================================================================================================================================================================================================
Installing:
 vim-enhanced                                                       x86_64                                                     2:8.2.3404-1.fc35                                                       fedora                                                     1.8 M
Installing dependencies:
 gpm-libs                                                           x86_64                                                     1.20.7-28.fc35                                                          fedora                                                      20 k
 libsodium                                                          x86_64                                                     1.0.18-8.fc35                                                           fedora                                                     161 k
 vim-common                                                         x86_64                                                     2:8.2.3404-1.fc35                                                       fedora                                                     6.7 M
 vim-filesystem                                                     noarch                                                     2:8.2.3404-1.fc35                                                       fedora                                                      22 k
 which                                                              x86_64                                                     2.21-27.fc35                                                            fedora                                                      41 k

Transaction Summary
========================================================================================================================================================================================================================================================================
Install  6 Packages

Total download size: 8.8 M
Installed size: 35 M
Downloading Packages:
(1/6): gpm-libs-1.20.7-28.fc35.x86_64.rpm                                                                                                                                                                                                79 kB/s |  20 kB     00:00    
(2/6): libsodium-1.0.18-8.fc35.x86_64.rpm                                                                                                                                                                                               235 kB/s | 161 kB     00:00    
(3/6): vim-enhanced-8.2.3404-1.fc35.x86_64.rpm                                                                                                                                                                                          2.8 MB/s | 1.8 MB     00:00    
(4/6): vim-filesystem-8.2.3404-1.fc35.noarch.rpm                                                                                                                                                                                         91 kB/s |  22 kB     00:00    
(5/6): which-2.21-27.fc35.x86_64.rpm                                                                                                                                                                                                    120 kB/s |  41 kB     00:00    
(6/6): vim-common-8.2.3404-1.fc35.x86_64.rpm                                                                                                                                                                                            3.3 MB/s | 6.7 MB     00:02    
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                                                                   4.4 MB/s | 8.8 MB     00:02     
Running transaction check
Transaction check succeeded.
Running transaction test
The downloaded packages were saved in cache until the next successful transaction.
You can remove cached packages by executing 'dnf clean packages'.
Error: Transaction test error:
  file /usr/share/vim/vim82/defaults.vim from install of vim-common-2:8.2.3404-1.fc35.x86_64 conflicts with file from package vim-minimal-2:8.2.3642-1.fc35.x86_64

# yum list installed | grep vi
vim-minimal.x86_64                       2:8.2.3642-1.fc35          @koji-override-1
```
</details>

### Note
We could experiment with setting curl retry options as well instead of hard-coding mirrors. That could help with Centos 7 issues as well.


~~Some mirrors (e.g. Halifax RWTH) repositories `repodata` seems out-of-sync. I have to double-check this, but the same repository settings wroks with some mirrors (where I see synced `repodata`).~~
Update: mirrors indeed were out-of-sync, and the issue is resolved.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by gaborznagy** [comment]
I've rerun my tests today on 3 mirrors I've tested against yesterday. All 3 of them were working today, and I see new files on the mirrors which didn't work yesterday, so it proves my theory about out-of-sync issues.

Here is the script I've used:

```
#!/usr/bin/env bash

DNF_INSTALL="dnf install -y"

#MIRROR_URL='https://ftp.halifax.rwth-aachen.de/fedora/linux'
#MIRROR_URL='https://mirror.szerverem.hu/fedora/linux'
MIRROR_URL='https://linuxsoft.cern.ch/fedora/linux'
# Workaround for often getting 503 from mirrors.fedoraproject.org.
sed -i -e "/baseurl/c\baseurl=${MIRROR_URL}/releases/\$releasever/Everything/\$basearch/os/" /etc/yum.repos.d/fedora.repo
sed -i -e "/baseurl/c\baseurl=${MIRROR_URL}/releases/\$releasever/Modular/\$basearch/os/" /etc/yum.repos.d/fedora-modular.repo
sed -i -e "/baseurl/c\baseurl=${MIRROR_URL}/updates/\$releasever/Everything/\$basearch/" /etc/yum.repos.d/fedora-updates.repo
sed -i -e "/baseurl/c\baseurl=${MIRROR_URL}/updates/\$releasever/Modular/\$basearch/" /etc/yum.repos.d/fedora-updates-modular.repo
# we don't need h264 codecs
sed -i -e "/enabled/c\enabled=0" /etc/yum.repos.d/fedora-cisco-openh264.repo

# testing changes
sed -i  -e '/metalink/s/mirrors\.fedora/tukrok.kalap/' /etc/yum.repos.d/*.repo
dnf clean all
rm -rf /var/cache/dnf/*
$DNF_INSTALL --verbose --best coreutils shadow-utils
```

And some log reference for the future about what did I see when a mirror is out-of sync:

(I've split the output into multiple lines for better readability)
```
Errors during downloading metadata for repository 'updates':
  - Status code: 404 for https://linuxsoft.cern.ch/fedora/linux/updates/35/Everything/x86_64/repodata/20417b414b0428c31677064afa586dd311fc0416e237ca8bccd591a1a077340e-primary.xml.zck (IP: 188.184.99.119)
  - Curl error (6): Couldn't resolve host name for https://tukrok.kalapproject.org/metalink?repo=updates-released-f35&arch=x86_64 [Could not resolve host: tukrok.kalapproject.org]
  - Status code: 404 for https://linuxsoft.cern.ch/fedora/linux/updates/35/Everything/x86_64/repodata/49f5d49e9678909d97da1495ed816582da265c6ab3ad15e51800489f41b88e2b-filelists.xml.zck (IP: 188.184.99.119)
Error: Failed to download metadata for repo 'updates': Yum repo downloading error: Downloading error(s):  
 repodata/20417b414b0428c31677064afa586dd311fc0416e237ca8bccd591a1a077340e-primary.xml.zck - Cannot download, all mirrors were already tried without success;
 repodata/49f5d49e9678909d97da1495ed816582da265c6ab3ad15e51800489f41b88e2b-filelists.xml.zck - Cannot download, all mirrors were already tried without success;
 repodata/189fca56d0a2e752ee9e88ed3092425e53fd961b455a2040951844e6ee445567-prestodelta.xml.zck - Cannot download, all mirrors were already tried without success;
 repodata/52f036de5cb0bfbcc62d09a22f7a1749bdad04d85bac2b105cf1e3033eb82076-updateinfo.xml.zck - Cannot download, all mirrors were already tried without success
```

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]
I've checked the metalink-to-baseurl fallback with a 5XX error code too, it works as expected.

Note: The fallback mechanism is probably not documented, at least I could not find it.

File: dbld/builddeps [hunk 1]
```diff
@@ -11,10 +11,16 @@ set -e
 set -x
 
 function workaround_rpm_repos() {
+    MIRROR_URL='https://ftp.halifax.rwth-aachen.de/fedora/linux/'
```
Comment (MrAnno) [inline dbld/builddeps:14]: I'd probably remove the trailing slash as it will be duplicated after the below sed commands.

Comment (gaborznagy) [reply dbld/builddeps]: I've removed the trailing slash.

**Comment by gaborznagy** [comment]
@MrAnno  Kindly re-approve. We are in dire need to merge both this PR and #3933.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: dbld/builddeps [hunk 2]
```diff
@@ -43,6 +43,7 @@ function install_dbld_dependencies()
             fi
             ;;
         fedora)
+            $DNF_INSTALL -y wget dnf-plugins-core
```
Comment (bazsi) [inline dbld/builddeps:46]: but shouldn't we remove them from add_copr_repo() at the same time?

Comment (gaborznagy) [reply dbld/builddeps:46]: Nope, `dnf builddep` is from `dnf-plugins-core`, if we would remove these packages then `install_rpm_build_deps` wouldn't work.

I guess `virtualenv` is the answer, or at least install packages per-user  and not system-wide.

**Comment by bazsi** [comment]
I think we could merge #3933 into this branch so that we can merge it in a green state and without having to reenable Fedora-33 builds in copr.

I am bulding this two merged to a single branch locally at the moment.

**Comment by bazsi** [comment]
The combination of the two branches was built properly for me. Just cherry-pick my patches on top of this one.

**Review by bazsi — DISMISSED** [review]

**Comment by gaborznagy** [comment]
> The combination of the two branches was built properly for me. Just cherry-pick my patches on top of this one.

Thanks!!! \o/ Hopefully this can be merged in soon.

**Comment by gaborznagy** [comment]
I've dropped the vim-workaround [patch](4a3b773f34033a03be52afcf8e97758f58a0bf9f) from #3933 .

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Pull request merged**